### PR TITLE
add lstm model

### PR DESCRIPTION
**Pull request by maypink**

### CODE REVIEW

File: models/NER/models/lstm/functions.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import numpy as np
+import torch.nn as nn
+from sklearn.metrics import precision_recall_curve
+
+
+def create_ris_dict():
```
Comment (ARQtty) [inline models/NER/models/lstm/functions.py:6]: docstring бы и типы возвращаемых аргументов

Comment (ARQtty) [reply models/NER/models/lstm/functions.py:6]: для всех  функций, непонятных за 3 секунды

File: models/NER/models/lstm/functions.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    a = ord('а')
+    chars = [chr(i) for i in range(a, a + 32)]
+    nums = [i + 1 for i in range(0, 33)]
```
Comment (ARQtty) [inline models/NER/models/lstm/functions.py:9]: мб `list(range(1, 33))` ?

File: models/NER/models/lstm/functions.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    rus_dict = dict(zip(chars, nums))
+    return rus_dict
+
+
+def encode(word, max_len):
```
Comment (ARQtty) [inline models/NER/models/lstm/functions.py:14]: мб сделать это с помощью embedding(word) и torch padding'a?

File: models/NER/models/lstm/functions.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        cur_word.append(rus_dict[i])
+    while len(cur_word) < max_len:
+        cur_word.append(1)
+    return cur_word
+
+
+def f_score(preds, labels):
+    sigmoid = nn.Sigmoid()
+    assert all(map(lambda preds: preds.size(1) == 2, preds))
+    def to_probs(preds): sigmoid(preds)[:, 0].cpu().detach().numpy()
```
Comment (ARQtty) [inline models/NER/models/lstm/functions.py:27]: это страшнее, чем добавить линтовую ошибку про лямбды в игнор линтера

File: models/NER/models/lstm/functions.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    def to_labels(labels): (1 - labels).cpu().detach().numpy()
+
+    preds = list(map(to_probs, preds))
+    labels = list(map(to_labels, labels))
+
+    preds = [item for subl in preds for item in subl]
```
Comment (ARQtty) [inline models/NER/models/lstm/functions.py:33]: возможно `itertools.chain` или `from_chain` - это то что тебе нужно здесь

File: models/NER/models/lstm/functions.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    labels = [item for subl in labels for item in subl]
+    precision, recall, thresholds = precision_recall_curve(labels, preds)
+    fscore = (2 * precision * recall) / (precision + recall)
+    ix = np.argmax(fscore)
```
Comment (ARQtty) [inline models/NER/models/lstm/functions.py:37]: Если входные данные - тензоры (стоит указать типы аргументов функции), то мб полностью торч методами посчитать f-score?  Каст к np, из-за которого потом нужно импортировать np для вызова argmax выглядит накручиванием сложностей

File: models/NER/models/lstm/model.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import torch
+import torch.nn as nn
+
+
+class LSTM_fixed_len(nn.Module):
```
Comment (ARQtty) [inline models/NER/models/lstm/model.py:5]: имена классов - camelCase. Странно, что линтер не заметил

File: models/NER/models/lstm/model.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        lstm_out, (ht, ct) = self.lstm(x)
+        last_ht = ht[-1]
+
+        y = self.fc_block(last_ht)
+
+        y = self.fc_block2(y)
+
+        y = self.fc(y)
+
```
Comment (ARQtty) [inline models/NER/models/lstm/model.py:41]: Ворд -> формат -> абзац -> межстрочный интервал -> 1.0?

File: models/NER/models/lstm/model.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    def encode(self, word, max_len=14):
+        cur_word = []
+        a = ord('а')
+        chars = [chr(i) for i in range(a, a + 32)]
+        nums = [i + 1 for i in range(0, 33)]
+        rus_dict = dict(zip(chars, nums))
+        for i in word.lower():
+            cur_word.append(rus_dict[i])
+        while len(cur_word) < max_len:
```
Comment (ARQtty) [inline models/NER/models/lstm/model.py:52]: А если больше?

File: models/NER/models/lstm/model.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            cur_word.append(1)
+        return torch.tensor(cur_word)
+
+    def prediction(self, word):
+        encoded = self.encode(word)
+        encoded = torch.reshape(encoded, (1, 14))
```
Comment (ARQtty) [inline models/NER/models/lstm/model.py:58]: Again а если слово длиннее?

File: models/NER/models/lstm/model.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        probability = self.forward(encoded)
+        return probability
+
+    def save(self, path):
```
Comment (ARQtty) [inline models/NER/models/lstm/model.py:62]: Раз save - метод класса, то стоит сделать и load методом

File: models/NER/models/lstm/token_dataset.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+class TokensDataset(Dataset):
+    def __init__(self, X, Y):
+        self.X = X
+        self.y = Y
+
+    def __len__(self):
+        return len(self.y)
+
+    def __getitem__(self, idx):
```
Comment (ARQtty) [inline models/NER/models/lstm/token_dataset.py:12]: В целом делать one-hot encoding можно и здесь. Вся логика будет инкапсулирована в датасет. Так будет правильнее

File: models/NER/models/lstm/token_dataset.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
+from torch.utils.data import Dataset
+
+
+class TokensDataset(Dataset):
```
Comment (ARQtty) [inline models/NER/models/lstm/token_dataset.py:4]: тут бы ещё collate_fn определить

**Review by ARQtty — CHANGES_REQUESTED** [review]

File: models/NER/models/lstm/model.py (hunk 8 above)
Comment (maypink) [reply models/NER/models/lstm/model.py:41]: image
поменяла на две, линтер кидает ошибку

File: models/NER/models/lstm/functions.py (hunk 2 above)
Comment (maypink) [reply models/NER/models/lstm/functions.py:9]: image

File: models/NER/models/lstm/functions.py (hunk 4 above)
Comment (maypink) [reply models/NER/models/lstm/functions.py:27]: твой же код взятый с auc из train_rubert

File: models/NER/models/lstm/functions.py (hunk 2 above)
Comment (ARQtty) [reply models/NER/models/lstm/functions.py:9]: не, коммент к 9ой строке

File: models/NER/models/lstm/model.py (hunk 8 above)
Comment (ARQtty) [reply models/NER/models/lstm/model.py:41]: Не, речь о том, что здесь вроде как между строками сильной логической разницы нет, поэтому нет смысла после каждой пустую ставить

File: models/NER/models/lstm/model.py (hunk 7 above)
Comment (maypink) [reply models/NER/models/lstm/model.py:5]: +

File: models/NER/models/lstm/model.py (hunk 9 above)
Comment (maypink) [reply models/NER/models/lstm/model.py:52]: +

File: models/NER/models/lstm/model.py (hunk 10 above)
Comment (maypink) [reply models/NER/models/lstm/model.py:58]: +

File: models/NER/models/lstm/model.py (hunk 11 above)
Comment (maypink) [reply models/NER/models/lstm/model.py:62]: +

File: models/NER/models/lstm/functions.py (hunk 5 above)
Comment (maypink) [reply models/NER/models/lstm/functions.py:33]: это опять кусок твоего кода...))))

**Review by ARQtty — APPROVED** [review]
